<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title></title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:root@bigsura.local" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a>
    <ul>
      <li><a href="#VERSION-v1.30">VERSION v1.30</a></li>
    </ul>
  </li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#Methods">Methods</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#set-methods-for-properties">set methods for properties</a></li>
      <li><a href="#write_using">write_using</a></li>
      <li><a href="#get_cover_batch">get_cover_batch</a></li>
      <li><a href="#readtr">readtr</a></li>
      <li><a href="#load_anno">load_anno</a></li>
      <li><a href="#region_merge">region_merge</a></li>
      <li><a href="#anno">anno</a></li>
      <li><a href="#varanno">varanno</a></li>
      <li><a href="#P1toP3">P1toP3</a></li>
      <li><a href="#finaliseAnno">finaliseAnno</a></li>
      <li><a href="#decide_major">decide_major</a></li>
      <li><a href="#getTrChange">getTrChange</a></li>
      <li><a href="#prWalker">prWalker</a></li>
      <li><a href="#trWalker">trWalker</a></li>
      <li><a href="#cmpPos">cmpPos</a></li>
      <li><a href="#getCPosFrame_by_cdsPos">_getCPosFrame_by_cdsPos</a></li>
      <li><a href="#cdsubstr">_cdsubstr</a></li>
      <li><a href="#getPairedCodon">_getPairedCodon</a></li>
      <li><a href="#translate">translate</a></li>
      <li><a href="#getTrRef">getTrRef</a></li>
      <li><a href="#batch_anno">batch_anno</a></li>
    </ul>
  </li>
  <li><a href="#BedAnno::Var">BedAnno::Var</a></li>
  <li><a href="#METHOD">METHOD</a>
    <ul>
      <li><a href="#new1">new</a></li>
      <li><a href="#getUnifiedVar">getUnifiedVar</a></li>
      <li><a href="#isSameTo">isSameTo</a></li>
      <li><a href="#parse_complex">parse_complex</a></li>
      <li><a href="#guess_type">guess_type</a></li>
      <li><a href="#get_internal">get_internal</a></li>
      <li><a href="#get_gHGVS">get_gHGVS</a></li>
    </ul>
  </li>
  <li><a href="#BedAnno::Anno">BedAnno::Anno</a></li>
  <li><a href="#METHOD1">METHOD</a>
    <ul>
      <li><a href="#new2">new</a></li>
      <li><a href="#getTrPosition">getTrPosition</a></li>
      <li><a href="#cal_hgvs_pos">cal_hgvs_pos</a></li>
    </ul>
  </li>
  <li><a href="#BedAnno::CNV">BedAnno::CNV</a></li>
  <li><a href="#METHOD2">METHOD</a>
    <ul>
      <li><a href="#new3">new</a></li>
      <li><a href="#annoCNV">annoCNV</a></li>
      <li><a href="#batch_annoCNV">batch_annoCNV</a></li>
    </ul>
  </li>
  <li><a href="#DATABASE-FORMAT">DATABASE FORMAT</a>
    <ul>
      <li><a href="#BED-FORMAT-DATABASE">BED FORMAT DATABASE</a></li>
      <li><a href="#TRANSCRIPT-FASTA-DATABASE">TRANSCRIPT FASTA DATABASE</a></li>
      <li><a href="#PRIMARY-TAG-ASSIGNMENT">PRIMARY TAG ASSIGNMENT</a></li>
    </ul>
  </li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#COPYRIGHT-AND-LICENSE">COPYRIGHT AND LICENSE</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>BedAnno - Perl module for annotating variation depend on the BED format database.</p>

<h2 id="VERSION-v1.30">VERSION v1.30</h2>

<p>From version 0.32 BedAnno will change to support CG&#39;s variant shell list and use ncbi annotation release 104 as the annotation database with reformatted database format, and won&#39;t give any individual annotation, so the <code>individual_anno</code> is no longer available. VCF4.1 format variant description (chr, pos, ref, alt) will also be supported.</p>

<p>From version v1.30 BedAnno will change to NOT to give extra annotations, no other plugins is supported and will completely clean, only for chromosome variant&#39;s gHGVS cHGVS pHGVS annotation and function predictions.</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code>  use BedAnno;
  my $beda = BedAnno-&gt;new( db =&gt; &quot;in.bed.gz&quot;, tr =&gt; &#39;in.trans.fas&#39; );
  my $anno = $beda-&gt;anno( &#39;chr20&#39;, 1234567, 1234569, &#39;AG&#39;, &#39;TGGG&#39; );</code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>By using this module, we can get variants from whole-genome or exome-capture NGS genotyping result annotated. The information contains various possible HGVS string together with a most recent strandard HGVS mutation name. It can not annotate ambiguous variants (transition, transvertion, or unknown break point large deletion and duplication).</p>

<p><i>BedAnno</i> annotate genomics variations of hg19 by using a BED format database, which construct from ncbi anno release 104, combined with tabix index. This module can directly parse the vcf4.1 format ref and single alt string (no commas in it), without normalized by vcftools, and can recognize the tandom repeat variation and duplication, generate the standard HGVS strings for most of complex cases. Also it will ajust the strand of transcript, and follow the 3&#39; nearest rules to annotate.</p>

<h1 id="Methods">Methods</h1>

<h2 id="new">new</h2>

<dl>

<dt id="About-:-Creat-a-new-annotation-entry">About : Creat a new annotation entry</dt>
<dd>

</dd>
<dt id="Usage">Usage :</dt>
<dd>

<pre><code>    my $beda = BedAnno-&gt;new( db =&gt; &quot;in.bed.gz&quot;, tr =&gt; &#39;in.trans.fas.gz&#39;, batch =&gt; 1 );</code></pre>

</dd>
<dt id="Args---all-database-files-should-be-tabix-indexed">Args - (all database files should be tabix indexed)</dt>
<dd>

<dl>

<dt id="Essential-Args">Essential Args:</dt>
<dd>

<dl>

<dt id="db-in.bed.gz"><i>db</i> &lt;in.bed.gz&gt;</dt>
<dd>

<dl>

<dt id="annotation-database-file">annotation database file.</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="tr-in.trans.fas.gz"><i>tr</i> &lt;in.trans.fas.gz&gt;</dt>
<dd>

<dl>

<dt id="transcript-sequence-fasta-file">transcript sequence fasta file</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="See-DATABASE-FORMAT-for-more-infomation">See <a href="#DATABASE-FORMAT">&quot;DATABASE FORMAT&quot;</a> for more infomation.</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="Optional-Args">Optional Args :</dt>
<dd>

<dl>

<dt id="Common-options">Common options :</dt>
<dd>

<dl>

<dt id="quiet"><i>quiet</i></dt>
<dd>

<dl>

<dt id="Suppress-warning-messege-to-output">Suppress warning messege to output.</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="batch-boolean"><i>batch</i> [boolean]</dt>
<dd>

<dl>

<dt id="use-batch-mode-annotation-default-in-daemon-mode-as-an-annotation-engine">use batch mode annotation, default in daemon mode as an annotation engine.</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="genome-refgenome.fa.gz"><i>genome</i> [ &quot;refgenome.fa.gz&quot; ]</dt>
<dd>

<dl>

<dt id="reference-genome-fasta-bgzipped-and-samtools-faidxed-for-use">reference genome fasta, bgzipped and samtools faidxed for use.</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="genes-genes.list-geneslist_filehandle"><i>genes</i> [ &quot;genes.list&quot; | geneslist_filehandle ]</dt>
<dd>

<dl>

<dt id="annotate-transcripts-for-genes.-e.g.-ABC-1-DEF-1-or-genes.list">annotate transcripts for <i>genes</i>. e.g. {&quot;ABC&quot; =&gt; 1, &quot;DEF&quot; =&gt; 1} or &quot;genes.list&quot;</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="trans-trans.list-translist_filehandle"><i>trans</i> [ &quot;trans.list&quot; | translist_filehandle ]</dt>
<dd>

<dl>

<dt id="annotate-transcripts-in-trans.-e.g.-NM_0012.1-1-NM_0034.2-1-or-trans.list">annotate transcripts in <i>trans</i>. e.g. {&quot;NM_0012.1&quot; =&gt; 1, &quot;NM_0034.2&quot; =&gt; 1} or &quot;trans.list&quot;</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="mmap-boolean"><i>mmap</i> [boolean]</dt>
<dd>

<dl>

<dt id="allow-annotating-all-other-non-BEST-multiple-mapping-records-boolen-option-default-not-allowed.-e.g.-NM_0123.1-have-3-mapping-location-or-alternate-splicing-default-only-the-BEST-one-will-be-annotated.-See-DATABASE-FORMAT-for-the-BEST-definition">allow annotating all other non &quot;BEST&quot; multiple-mapping records, boolen option, default not allowed. e.g. NM_0123.1 have 3 mapping location or alternate splicing, default only the &quot;BEST&quot; one will be annotated. See <a href="#DATABASE-FORMAT">&quot;DATABASE FORMAT&quot;</a> for the &quot;BEST&quot; definition.</dt>
<dd>

</dd>
</dl>

</dd>
</dl>

</dd>
<dt id="Batch-mode-options">Batch mode options :</dt>
<dd>

<dl>

<dt id="region-region_string"><i>region</i> [region_string]</dt>
<dd>

<dl>

<dt id="limit-to-only-annotate-transcript-in-region.-e.g.-chr20:1234567-1234568-prior-to-regbed">limit to only annotate transcript in <i>region</i>. e.g. &quot;chr20:1234567-1234568&quot;, prior to <i>regbed</i>.</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="regbed-BED-format-file"><i>regbed</i> [BED format file]</dt>
<dd>

<dl>

<dt id="similar-to-region-with-allowing-multiple-regions.-e.g.-in.region.bed">similar to <i>region</i>, with allowing multiple regions. e.g. &quot;in.region.bed&quot;.</dt>
<dd>

</dd>
</dl>

</dd>
</dl>

</dd>
<dt id="Notes">Notes</dt>
<dd>

<dl>

<dt id="Batch-mode-is-designed-for-annotation-of-a-complete-list-of-variants-on-same-chromosome-read-all-information-of-the-chr-into-memory-and-annotate-all-variants-together-in-the-order-of-chr-coordinates.-This-mode-can-avoid-frequent-IO-brought-by-tabix-searching-but-need-huge-memory-cost">Batch mode is designed for annotation of a complete list of variants on same chromosome, read all information of the chr into memory, and annotate all variants together in the order of chr coordinates. This mode can avoid frequent IO, brought by tabix searching, but need huge memory cost.</dt>
<dd>

</dd>
</dl>

</dd>
</dl>

</dd>
</dl>

</dd>
<dt id="Returns">Returns</dt>
<dd>

<dl>

<dt id="Annotation-Engine-object-entry-please-see-load_anno-for-more-information">Annotation Engine object entry, please see <a href="#load_anno">&quot;load_anno&quot;</a> for more information.</dt>
<dd>

</dd>
</dl>

</dd>
</dl>

<h2 id="set-methods-for-properties">set methods for properties</h2>

<pre><code>    List of Properties:
                        set
    db                  o
    tr                  o
    refbuild            o
    genome              o

    e.g.    : $beda-&gt;set_refbuild($refbuild);</code></pre>

<h2 id="write_using">write_using</h2>

<pre><code>    About   : write the current using database information to files
    Usage   : $beda-&gt;write_using( $file, $type );
    Args    : file gives the filename to output, and type is one of the following:
                g:  gene symbol list
                t:  transcript acc.ver list
                c:  the complete annotation region, in bed format,
                    which can be used as the variation calling region.
                b:  standard bed format of only exon region
                a:  BED format, exon region with &#39;+1&#39; annotation, 
                    oneline per one exon for one transcript, this
                    format allow redundancy exists, and not sorted by
                    chromosomal coordinates, but by the transcript acc.ver
                    and the exon number, this file is mainly for 
                    transcript originated statistics for only exon.</code></pre>

<h2 id="get_cover_batch">get_cover_batch</h2>

<pre><code>    About   : get covered region in batch mode
    Usage   : my $cover_href = $beda-&gt;get_cover_batch( $chr, \@stasto );
    Args    : a chromosome id, and an array ref of [ [ $start, $stop ], ... ]
    Returns : a hash ref of:
                {
                    &quot;$start-$stop&quot; =&gt; [ 
                                        [ $tid, $left_cPos, $right_cPos ], ... 
                                      ], ...
                    # start is 1 based, to coordinate with region string format
                }
              Note: the pos-pair which do not hit any annotation blocks, will
                    not exist in the returned results.</code></pre>

<h2 id="readtr">readtr</h2>

<pre><code>    About   : Read transcript information, and even sequences if in batch mode.
    Usage   : my $rtrSeqs = $beda-&gt;readtr( genes =&gt; $rh_NewGenes, trans =&gt; $rh_NewTrans );
    Args    : Optional args &quot;genes&quot; and &quot;trans&quot; only accept hash ref values.
              if no args specified, it will load information based on the
              configuration of BedAnno entry.
    Returns : A hash ref of trSeqs:
              {
                $tr_acc =&gt; {
                    len      =&gt; $tr_len,
                    gene     =&gt; $gene_sym,

                    # optional tags:
                    prot     =&gt; $prot_acc,
                    plen     =&gt; $prot_len,
                    csta     =&gt; $cds_start_on_trSeq, # 0 based
                    csto     =&gt; $cds_end_on_trSeq,   # 1 based
                    seq      =&gt; $tr_sequence,
                    
                    nfs      =&gt; { $fslead =&gt; $fsbase, ... },
                    cfs      =&gt; { $cfslead =&gt; $fsbase, ... },
                    X        =&gt; 1,                   # inseqStop
                    U        =&gt; 1,                   # selenocysteine
                    A        =&gt; 1,                   # polyATail
                    altstart =&gt; {                    # altstart codons
                        $startCodons1 =&gt; 1,
                        $startCodons2 =&gt; 1,
                        ...
                    },

                    # the following two keys will be assigned
                    # to mRNA when needed
                    cseq     =&gt; $codonSequence,
                    pseq     =&gt; $proteinSequence,
                },
                ...
              }</code></pre>

<h2 id="load_anno">load_anno</h2>

<pre><code>    About   : load all needed annotation infomation into memory for multi-process annotation
    Usage   : my $ranndb = $beda-&gt;load_anno( region =&gt; &quot;chr20:1234567-1234568&quot;, trans =&gt; \%trans );
    Args    : Using %args to override class&#39;s properties: region, regbed, genes, trans
              if no args, use the the class&#39;s properties as default.
    Returns : a localized merged anno db, The returned annotation database is a hash ref.
        {
            $chr =&gt; [
                {
                    sta   =&gt; $start, (0 based)
                    sto   =&gt; $stop,  (1 based)
                    annos =&gt; {
                        $anno_string =&gt; $offset, ...
                    }

                    detail =&gt; {
                        $tid =&gt; {
                            gsym =&gt; $gsym,    (gene symbol)
                            gid  =&gt; $gid,     (Entrez gene id)
                            gpSO =&gt; $gpSO,    (GeneParts SO)
                            blka =&gt; $blka,    (block attribute)
                            exin =&gt; $exin,    (exon intron number)
                            nsta =&gt; $nsta,    (n./r. left  of whole block)
                            nsto =&gt; $nsto,    (n./r. right of whole block)
                            csta =&gt; $csta,    (c.    left  of whole block)
                            csto =&gt; $csto,    (c.    right of whole block)
                            wlen =&gt; $wlen,    (length of whole block)
                            pr   =&gt; $pr,      (primary tag)
                            strd =&gt; $strd,    (strand)
                            offset =&gt; $offset,(offset of current block to whole block)
                            mismatch =&gt; $mismatch (non-equal length block descripter)
                        }, ...
                    }
                }, ... 
            ], ...
        }
      Note: when variation hit one of the annotation entry, the anno_string will be parsed.
      and the &quot;detail&quot; tag will be added then.</code></pre>

<h2 id="region_merge">region_merge</h2>

<pre><code>    About   : merge consecutive same-entries regions
    Usage   : my $rannodb = region_merge($loaded_db);
    Args    : A hash ref of loaded_db.
    Returns : A hash ref of merged db.</code></pre>

<h2 id="anno">anno</h2>

<pre><code>    About   : Annotate single short variation by annotation db.
    Usage   : my $anno_ent = $beda-&gt;anno( &#39;chr20&#39;, 1234567, 1234569, &#39;AG&#39;, &#39;TGGG&#39; );
              or $anno_ent = $beda-&gt;anno( &#39;chr20&#39;, 1234568, &#39;AG&#39;, &#39;AGGG&#39; );
    Args    : for CG&#39;s shell variants, need 5 args in UCSC coordinates
              (0-based start), they are:
                chr id, chr start, chr end, reference, alternative.
              for variants in VCF, need 4 args, which is lack of 
                chr end, and &quot;chr start&quot; is in 1-based coordinates.
              for crawler: a input object with keys: 
                chr,begin,referenceSequence,variantSequence,[end].
                if end is specified, then use 0-based coordinates,
                otherwise 1-based (VCF) coordinates will be used.
    Returns : a hash ref of annotation informations, see varanno().</code></pre>

<h2 id="varanno">varanno</h2>

<pre><code>    About   : implicitly create a new BedAnno::Anno entry, and 
              assign all the needed annotation for to it.
    Usage   : ($rAnnoRst, $AEIndex) = $beda-&gt;varanno($var, $AEIndex);
    Args    : The BedAnno entry, BedAnno::Var entry and current dbidx, 
              current dbidx should be always 0, if used for non-batch mode.
    Returns : A BedAnno::Anno entry and current dbidx for nex query in batch.
            {
                var =&gt; {

                    # the first part is from var parsing result.
                    # please see &quot;BedAnno::Var new()&quot;.
                    # import all the keys from original parsed var entry
                    # and add the following keys by this method.

                    varName =&gt; $var_mutation_name,

                    # information
                    varTypeSO =&gt; $varTypeSO,
                    gHGVS     =&gt; $gHGVS,
                    refbuild  =&gt; $referenceBuild,
                },
                trInfo =&gt; {
                    $tid =&gt; {
                        trVarName     =&gt; $transcriptVariantName,
                        geneId        =&gt; $Entrez_Gene_ID,
                        geneSym       =&gt; $Gene_Symbol,
                        prot          =&gt; $Protein_Acc_Ver,
                        strd          =&gt; $strand,
                        rnaBegin      =&gt; $Begin_in_RNA_transcript,
                        rnaEnd        =&gt; $End_in_RNA_transcript,
                        cdsBegin      =&gt; $Begin_in_CDS,            # cDot format
                        cdsEnd        =&gt; $End_in_CDS,              # cDot format
                        protBegin     =&gt; $Begin_in_Protein,
                        protEnd       =&gt; $End_in_Protein,
                        c             =&gt; $cHGVS,
                        p             =&gt; $pHGVS,
                        p3                =&gt; $threeletter_pHGVS,
                        cc            =&gt; $codon_change,
                        polar         =&gt; $polar_change,
                        r             =&gt; $imp_funcRegion,
                        r_Begin       =&gt; $imp_beginfuncRegion,
                        r_End         =&gt; $imp_endfuncRegion,
                        func          =&gt; $imp_funcCode,
                        exin          =&gt; $exIntr_number,
                        ei_Begin      =&gt; $imp_Begin_exIntr_number,
                        ei_End        =&gt; $imp_End_exIntr_number,
                        genepart      =&gt; $GenePart,
                        genepartSO    =&gt; $GenePartSO,
                        componentIndex =&gt; $componentIndex,
                        exonIndex     =&gt; $exonIndex,               # &#39;.&#39; for N/A
                        intronIndex   =&gt; $intronIndex,             # &#39;.&#39; for N/A
                        funcSOname    =&gt; $FunctionImpact,
                        funcSO        =&gt; $FunctionImpactSO,
                        trAlt         =&gt; $alt_string_on_transcript,
                        trRef         =&gt; $ref_string_on_transcript,
                        prAlt         =&gt; $protein_alt_sequence,
                        prRef         =&gt; $protein_ref_sequence,
                        primaryTag    =&gt; $refstandard_primary_or_not,   # Y/N
                        preStart =&gt; {    # the position before the start of var
                            nDot =&gt; $rna_hgvs_pos,
                            cDot =&gt; $cds_hgvs_pos,
                            r    =&gt; $func_region,
                            exin =&gt; $exon_intron_number,
                        },
                        postEnd =&gt; {     # the position after the end of var
                            nDot =&gt; $rna_hgvs_pos,
                            cDot =&gt; $cds_hgvs_pos,
                            r    =&gt; $func_region,
                            exin =&gt; $exon_intron_number,
                        },
                        trRefComp =&gt; {

                            # some trRef components
                        },

                        # for some of splice variants, there may exists 
                        # the following function information
                        alt_func        =&gt; $alternative_func_code,
                        alt_funcSO      =&gt; $alternative_variant_SO_id,
                        alt_funcSOname  =&gt; $alt_variant_SO_name,
                    },
                    ...
                }
            }</code></pre>

<h2 id="P1toP3">P1toP3</h2>

<pre><code>    About : Change 1 letter format of pHGVS string to 3 letter format
    Usage : my $p3 = P1toP3($p1);</code></pre>

<h2 id="finaliseAnno">finaliseAnno</h2>

<pre><code>    About   : finalise the BedAnno::Anno entry by check all tag values,
              and uniform them for AE output usage, query transcript
              oringinated additional resources and add them into the data
              frame.
    Usage   : $beda-&gt;finaliseAnno($annEnt);
    Args    : BedAnno entry and a BedAnno::Anno entry
    Returns : A finalised BedAnno::Anno entry</code></pre>

<h2 id="decide_major">decide_major</h2>

<pre><code>    About   : In finalise step, decide a major transcript to report for a var.
    Usage   : my $majorTranscriptVarName = decide_major($annoEnt);
    Returns : A string in the following format:
              If the transcript has a pName: 
                  mrnaAcc(geneSymbol): cName (pName), 
                  e.g. NM_145651.2(SCGB1C1): c.13C&gt;T (p.R5C)
              If the transcript does not have a pName: 
                  mrnaAcc(geneSymbol): cName
              If only intergenic
                  chr: gName (intergenic)
                  e.g. chrX: g.220025A&gt;T (intergenic)
    Note    : use the primaryTag to find reference standard or primary transcript,
              if only one have the primaryTag &quot;Y&quot;, then use it,
              otherwise, sort by GenePart: 
                1.  CDS
                2.  span
                3.  five_prime_cis_splice_site
                4.  three_prime_cis_splice_site
                5.  ncRNA
                6.  five_prime_UTR
                7.  three_prime_UTR
                8.  interior_intron
                9.  five_prime_UTR_intron
                10. three_prime_UTR_intron
                11. abnormal-intron
                12. promoter
                13. annotation-fail
                14. intergenic_region
              and choose the first one, if more than one transcript have the 
              same reference standard and same GenePart, then choose the first
              one which prior in name sorting.</code></pre>

<h2 id="getTrChange">getTrChange</h2>

<pre><code>    About   : Calculate the transcript changes, based on TrPostition
    Usage   : $beda-&gt;getTrChange($annoEnt);
    Returns : assign the following tags in annoEnt
                trRef, prot, c, p, cc, polar, func
                prRef, prAlt</code></pre>

<h2 id="prWalker">prWalker</h2>

<pre><code>    About   : act as trWalker, but on protein sequence
    Usage   : $prVar = $beda-&gt;prWalker( $prSeq,  $prBegin, $prEnd, $prRef, $prAlt );
    Args    : prSeq   - whole protein sequence
              prBegin - protein variant Begin position (1 based)
              prEnd   - protein variant End position (1 based)
              prRef   - protein variant reference sequence
              prAlt   - protein variant alt sequence</code></pre>

<h2 id="trWalker">trWalker</h2>

<pre><code>    About   : walk around the variant position to find possible
              repeat start/end, and return the recalculated
              trBegin and trEnd, together with the real
              transcript originated variants and unified property
              Current implementation won&#39;t walk around in the following
              cases:
              1. no-call
              2. annotation-fail
              3. snv or mnp
              4. non-exon region
              5. span case or any edge case
              6. delins without repeat.
    Usage   : ($trBegin, $trEnd, $real_var) = 
              $beda-&gt;trWalker($tid, $rtrinfo);
    Args    : trAcc and hash ref of trInfo annotation for trAcc,
          </code></pre>

<h2 id="cmpPos">cmpPos</h2>

<pre><code>    About   : judge the order for p1 and p2, because the insertion
              will have a reverted order of left &amp; right position
    Usage   : my $cmpRst = BedAnno-&gt;cmpPos($p1, $p2);
    Args    : hgvs positio p1 and p2, with out &#39;c.&#39; or &#39;n.&#39; flag
    Return  : 0 for same, 1 for normal order, -1 for reversed order.</code></pre>

<h2 id="getCPosFrame_by_cdsPos">_getCPosFrame_by_cdsPos</h2>

<pre><code>    About   : calculate codon position and frame info by
              involving frameshift case in consideration.
    Usage   : my @cPosFrame = _getCPosFrame_by_cdsPos( $trdbEnt, $cds_p );
    Args    : cds_p is relative position to the first bp in start codon on trans
    Returns : an array in the format of:
              ( $stat, $ra_posrefs1, $ra_posrefs2 )
              ra_posrefs* is an array ref for the following array
              [ $codonPos, $frame ]

              for non-cds region case stat is -1, without ra_posrefs.
              for normal case stat is 1, 
                  ra_posrefs1 is the information of input position,
                  without ra_posrefs2 specified.
              for deleted frame case stat is 0,
                  ra_posrefs1 is the information of position on 5&#39; 
                  side next to the deleted frame, and ra_posrefs2
                  is on 3&#39; side.
              for duplicated frame case stat is 2,
                  ra_posrefs1 is the 5&#39; most hit of input position,
                  ra_posrefs2 is the corresponding position on the 
                  3&#39; duplicated region.</code></pre>

<h2 id="cdsubstr">_cdsubstr</h2>

<pre><code>    About   : substr from transcript seqeunce involving frameshift changing.
    Usage   : my $codonStr = _cdsubstr( $trdbEnt, $start, $length, $replace );
    Args    : trdbEnt is a sub hash in trInfodb which contains csta, csto.
              start is 0 based start position of sub string on transcript seq.
              length is the length of target region on transcript seq.
              replace is the alternative.
    Returns : A substring cut from the transcript sequence, with frameshift involved.
              Behave like function &quot;substr&quot;, but for replace mode, 
              it returns the whole transcript seq after replacement.</code></pre>

<h2 id="getPairedCodon">_getPairedCodon</h2>

<pre><code>    About   : get codon position, codon string, aa string, and frame info
              for a pair of transcript position
    Usage   : my ($rcinfo5, $rcinfo3) = _getPairedCodon( $trdbEnt, $p5, $p3 );
    Args    : trdbEnt is a sub hash in trInfodb which contains csta, csto
              for cds start/end position, besides many other feature tags.
              p5 and p3 is a pair of positions which give a region or single
              position or maybe an insertion anchor.
    Returns : codon info array ref for p5 and p3. The array ref:
              [
                  AA position  - 0 for not in cds region.
                  codon string - codon string, e.g. &quot;ATA&quot;.
                  aa char      - AA code, 1 bp mode, e.g. &quot;E&quot;.
                  polar        - Polar properties, e.g. &quot;P+&quot;.
                  frame        - current position&#39;s frame info,
                                 -1 for not available or in fs-site.
                  [frame-alt]  - frame info around fs-site.
              ]</code></pre>

<h2 id="translate">translate</h2>

<pre><code>    About   : Translate nucleotides to aa seqs
    Usage   : my ($aa_seq, $next_frame) = translate( $nucl, { mito =&gt; 1, polyA =&gt; 1 } );
    Args    : first args should be the nucleotide seq to be translated,
              the second args is a hash ref of optional args (all boolean):
              mito   : indicate it is for mDNA (mitochondrion)
              nostop : indicate there&#39;s no stop codon in aa seq,
                       translate &#39;UGA&#39; to &#39;U&#39;, and other stop codon to &#39;X&#39;
              polyA  : indicate extra A should be added to 3&#39;end of codon,
                       to help encode a stop codon (usually used with mito).
    Returns : $aa_seq is the aa sequence, 
              $next_frame gives the next base&#39;s frame to the 3&#39;end of sequence.</code></pre>

<h2 id="getTrRef">getTrRef</h2>

<pre><code>    About   : generate concatenated transcript originated reference
    Usage   : my $trRef = getTrRef( $trannoEnt, $refgenome, $trSeq, $strd );
    Args    : trannoEnt - BedAnno::Anno-&gt;{trInfo}-&gt;{$tid}
              refgenome - Unified reference in BedAnno::Var
              trSeq     - whole transcript
              strd      - strand of transcript.
    Returns : transcript originated reference
    Notes   : Using sequence from transcript as the exon part,
              and using the sequence from reference genome
              as the intron part. and concatenate them.</code></pre>

<h2 id="batch_anno">batch_anno</h2>

<pre><code>    About   : The fastest way to annotate multiple snv and 1bp deletion variations,
              indel and other types also can be annotated, but no faster than annotated
              one by one.
    Usage   : $beda = BedAnno-&gt;new( db =&gt; &#39;in.bed.gz&#39;, tr =&gt; &#39;in.trans.fas&#39;, batch =&gt; 1);
              $rAnnoRst = $beda-&gt;batch_anno($rVars);
    Args    : an array ref of BedAnno::Var entries.
    Returns : an array ref of BedAnno::Anno entries, see varanno().</code></pre>

<h1 id="BedAnno::Var">BedAnno::Var</h1>

<pre><code>    BedAnno::Var sub module</code></pre>

<h1 id="METHOD">METHOD</h1>

<h2 id="new1">new</h2>

<pre><code>    About   : Create a new object class entry, BedAnno::Var,
              parse the variation directly by the ref and alt string.
    Usage   : my $var = BedAnno::Var-&gt;new( $chr, $start, $end, $ref, $alt );
           or my $var = BedAnno::Var-&gt;new( $chr, $pos, $ref, $alt );
           or my $var = BedAnno::Var-&gt;new( $varInput );
    Args    : Input can be variable format
              1. 5 parameters format: CG shell list format: chr,start,end,ref,alt
              2. 4 parameters format: VCF format: chr,pos,ref,alt
              3. Crawler input object: A hash ref with nessesary keys: 
                 chr,begin,referenceSequence,variantSequence,  
                 optional key is &quot;end&quot;, if end specified,
                 coordinates are treat as 0-based, otherwise, use 1-based (VCF)
    Returns : a new BedAnno::Var entry :
            {
                chr    =&gt; $chr,
                pos    =&gt; $start,          # 0-based start
                end    =&gt; $end,
                ref    =&gt; $ref,
                alt    =&gt; $alt,
                reflen =&gt; $ref_len,
                altlen =&gt; $alt_len,        # not exists if no-call
                guess  =&gt; $varType,        # the output varType
                imp    =&gt; $imp_varType,    # the implicit varType
                sm     =&gt; $sm,             # single/multiple base indicator
                                           # equal/non-equal length indicator

                # for hgvs naming convinient, reparse delins(guess),
                # in forward and reverse strand separately,
                # If the result are the same, then only
                # give the following optional
                # rescaled strand-same description group
                bp  =&gt; $bc_pos,       # backward compatible pos, 0-based
                br  =&gt; $bc_ref,       # backward compatible ref string
                ba  =&gt; $bc_alt,       # backward compatible alt string
                brl =&gt; $bc_reflen,    # backward compatible ref length
                bal =&gt; $bc_altlen,    # backward compatible alt length

                # otherwise, the following &#39;+&#39;, &#39;-&#39;,
                # structure will be generated to reflect
                # the difference. they are all optional

                &#39;+&#39; =&gt; {

                  # This group simplely trim off the leading same chars
                  # on forward strand, and then trim the same tail
                  bp  =&gt; $backward_fpos,
                  br  =&gt; $backward_fref,
                  ba  =&gt; $backward_falt,
                  brl =&gt; $backward_freflen,
                  bal =&gt; $backward_faltlen,

                },

                &#39;-&#39; =&gt; { 
                  # similar to &#39;+&#39;, but for reverse strand 
                },

                # this group gives ref/alt string based on the rule
                # with &#39;rep&#39; annotation available
                p      =&gt; $rep_left_pos,         # repeat pos, 0-based
                r      =&gt; $rep_ref,              # repeat ref string
                a      =&gt; $rep_alt,              # repeat alt string
                rl     =&gt; $rep_reflen,           # repeat ref length
                al     =&gt; $rep_altlen,           # repeat alt length
                rep    =&gt; $repeat_element,
                replen =&gt; $repeat_element_length,
                ref_cn =&gt; $copy_number_in_ref,
                alt_cn =&gt; $copy_number_in_alt,

                # for equal length long substitution
                # record the separated snvs positions
                # all positions are 1 based.
                sep_snvs =&gt; [ $snv_pos1, $snv_pos2, ... ],
            }</code></pre>

<h2 id="getUnifiedVar">getUnifiedVar</h2>

<pre><code>    About   : uniform the pos and ref/alt pair selection,
              after new(), give info for HGVS naming.
    Usage   : my @unified_desc = $var-&gt;getUnifiedVar($strd);
    Args    : BedAnno::Var entry and current strand for annotation.
    Returns : an array of ( 
                $pos,  # 0-based start pos
                $ref,  # reference bases
                $alt,  # called bases
                $reflen, # reference len
                $altlen )# called len, undef if no-call</code></pre>

<h2 id="isSameTo">isSameTo</h2>

<pre><code>    About   : Check if the current var is the same with another var
    Usage   : my $isSame = $var-&gt;isSameTo($another_var);
    Args    : BedAnno::Var entry.
    Returns : 1 for Same, 0 for NOT same.</code></pre>

<h2 id="parse_complex">parse_complex</h2>

<pre><code>    About   : parse complex delins variants to recognize 
              repeat and differ strand-pos var.
    Usage   : my $var = $var-&gt;parse_complex();
    Args    : variantion entry, which have been uniform to 
              CG&#39;s shell list format, with its &#39;guess&#39;:delins.
    Returns : see BedAnno::Var-&gt;new()</code></pre>

<h2 id="guess_type">guess_type</h2>

<pre><code>    About   : guess the varType directly from the input information.
    Usage   : my ($guess, $implicit_varType, $sm) = guess_type($len_ref, $ref, $alt);
    Args    : 1. length of reference (derived from end - start)
              2. reference sequence (&#39;.&#39; or empty for ins )
              3. called sequence ( &#39;?&#39; for no-call, &#39;.&#39; or empty for del )
    Returns : $guess is varType in output (ref,snv,ins,del,delins,no-call)
              $implicit_varType (ref,snv,ins,del,delins,rep)
              $sm is single/multiple/equal/non-equal-len indicator
                0 - ins case
                1 - single base ref case
                2 - multiple base, different length case
                3 - multiple base, equal length case </code></pre>

<h2 id="get_internal">get_internal</h2>

<pre><code>    About   : recalculate ref alt for delins and mutiple sample caused ref-alt pair.
              depend on different strand of the to-annotated gene or transcript, 
              the offset may be different for the same ref and alt,
              because of the 3&#39;end nearest annotation rules.
    Usage   : my $rephase = get_internal( $ref, $reflen, $alt, $altlen );
    Returns : a hash ref of : 
                {
                    &#39;+&#39; =&gt; $f_lofs,
                    &#39;-&#39; =&gt; $r_lofs,
                    &#39;r&#39; =&gt; $new_ref_len, 
                    &#39;a&#39; =&gt; $new_alt_len
                }</code></pre>

<h2 id="get_gHGVS">get_gHGVS</h2>

<pre><code>    About   : get genomic (chromosomal) HGVS string of variation
    Usage   : my $gHGVS = $var-&gt;get_gHGVS();
    Args    : variation entry, after BedAnno::Var-&gt;new().
    Returns : chromosomal HGVS string.</code></pre>

<h1 id="BedAnno::Anno">BedAnno::Anno</h1>

<pre><code>    BedAnno::Anno sub package</code></pre>

<h1 id="METHOD1">METHOD</h1>

<h2 id="new2">new</h2>

<pre><code>    About   : Create BedAnno::Anno object
    Usage   : my $annoEnt = BedAnno::Anno-&gt;new($var);
    Args    : BedAnno::Var entry
    Returns : BedAnno::Anno entry</code></pre>

<h2 id="getTrPosition">getTrPosition</h2>

<pre><code>    About   : Assign the BedAnno::Anno obj&#39;s trInfo with affected regions
    Usage   : my $AEIndex = $annoEnt-&gt;getTrPosition($rannodb, $AEIndex);
    Args    : $rannodb is a chromosome branch of BedAnno object&#39;s annodb feature,
              $AEIndex is the current index for annodb searching.
    Returns : A new AEIndex for next query.
    Notes   : $AEIndex is used for same chr batch mode.
              assign the following tag to $annoEnt
              {
                trInfo =&gt; {
                    $tid =&gt; {
                        geneId,   geneSym, strd, primaryTag,
                        trAlt =&gt; $stranded_alt_string_with_ext_at_mismatches,

                        preStart =&gt; {
                            nDot, cDot, exin, r
                        },

                        postEnd =&gt; {
                            nDot, cDot, exin, r
                        },

                        trRefComp =&gt; {
                            $exon_number   =&gt; $transcript_exon_length,
                            $intron_number =&gt; [
                                $start_in_non_stranded_reference,
                                $stop_in_non_stranded_reference
                              ],
                            ...
                        },
                    },
                    ...
                }
              }</code></pre>

<h2 id="cal_hgvs_pos">cal_hgvs_pos</h2>

<pre><code>    About   : calculate nDot, cDot HGVS position, depend on given offset,
              assign trAlt string and nDot HGVS and cDot HGVS positions.
    Usage   : $annoEnt-&gt;cal_hgvs_pos(
                    offset =&gt; $offset, 
                    tid    =&gt; $tid,
                    LR     =&gt; $lr,
                    tidDetail =&gt; $rh_tidDetail,
              );
    Args    : ofst is total offset to the start(left) of currunt annoblk,
              tid is the transcript id for the detail entry
              tidDetail is the currunt annoblk detail
              LR indicate this offset is left or right pos,
                1 for left and assign sta group,
                0 for right and assign end group.
              &quot;noassign&quot; to indicate whether to assign those information
              to annoEnt, it&#39;s used return the useful information only,
              without change the annoEnt. 
    Returns : if noassign is used, then return a hash ref, which contains
                { nDot, cDot, exin, r } if successful.
              otherwise, 0 for no assigned status, 1 for successful assigned.
             
    Notes   : For position mirror on transcript, there are 2 other cases 
              than normal case:
              1. annotation fail, which can not be annotated in the region
                 of it, the bad alignment string start with &#39;E&#39;.
              2. block with length changing mismatches, or long substitution
                 mismatch, which contain the following three cases:
                 
                 a. insertion (I) on refSeq

                        +-------+---+--------+  refgenome
                         \       \ /        /
                          +-------+--------+    refSeq

                 b. deletion (D) on refSeq 

                          +-------+--------+    refgenome
                         /       / \        \
                        +-------+---+--------+  refSeq

                 c. delins (S/DI) on refSeq (equal/non-equal length)

                        +-------+---+--------+  refgenome
                        |       |  /        /
                        +-------+-+--------+    refSeq

                 Insertion will have an reversed start/end position on refSeq,
                 due to the 1-based position description system.
                 
                 Any position located in a non-zero length refgenome mismatch
                 block have to extend to total region of mismatched block,
                 and alternate &quot;trAlt&quot; value in annotation entry for this tid.

              This method assign the following tag to $annoEnt
                {
                    trInfo =&gt; {
                        $tid =&gt; {
                            rnaBegin, rnaEnd,  cdsBegin, cdsEnd,
                            ei_Begin, ei_End,  r_Begin,  r_End,
                            genepartSO, trAlt,
                          },
                        ...
                    }
                }</code></pre>

<h1 id="BedAnno::CNV">BedAnno::CNV</h1>

<pre><code>    BedAnno::CNV sub package</code></pre>

<h1 id="METHOD2">METHOD</h1>

<h2 id="new3">new</h2>

<pre><code>    About   : Create BedAnno::CNV object
    Usage   : my $cnva = BedAnno::CNV-&gt;new( db =&gt; $annodb, tr =&gt; $trdb, dgv =&gt; $dgvdb, cnvPub =&gt; $cnvPubdb );
    Args    : required args: db, tr
              Specific args:
                - ovlp_rate   overlapping region rate while searching for hits
                - max_uncov   maximum uncovered region length in querys while searching.
              Same args in BedAnno method &#39;new&#39;:
                - db, tr, genes, trans, region, regbed, mmap, batch
    Returns : BedAnno::CNV object</code></pre>

<h2 id="annoCNV">annoCNV</h2>

<pre><code>    About   : Annotate single CNV variants
    Usage   : my $rcnvAnno = $cnva-&gt;annoCNV( $chr, $start, $end, $copy_number );
    Args    : $chr   - chromosome number
              $start - 0 based start position
              $end   - 1 based end position
              $copy_number - estimated copy number
    Returns : A hash ref of cnv annotation in the following structure
                {
                    cnva_type =&gt; $cnv_anno_type,

                    # if hit on transcript
                    anno =&gt; {
                        $tid =&gt; {
                            gsym   =&gt; $gene_symbol,
                            gid    =&gt; $gene_id,
                            strd   =&gt; $strand,
                            cpos   =&gt; $hgvs_position,
                            exin   =&gt; $exon_intron_number,
                            regcod =&gt; $code_of_region,
                            regtyp =&gt; $type_of_region,
                        },
                        ...
                    },
                }</code></pre>

<h2 id="batch_annoCNV">batch_annoCNV</h2>

<pre><code>    About   : Annotate CNV variants in batch mode
    Usage   : my $rcnvAnnos = $cnva-&gt;batch_annoCNV( $ref_hash_AllCNV );
    Args    : A hash ref of all cnv annotation in the following structure
                {
                    $chr =&gt; {
                        &quot;$start-$stop&quot; =&gt; $copy_number, 
                        # start is 1 based, to coordinate with region string format
                        ...
                      },
                      ...
                }
    Returns : A hash ref of annotated cnv variants in the following structure
                {
                    $chr =&gt; {
                        &quot;$start-$stop&quot; =&gt; {
                           cnva_type =&gt; $cnv_anno_type,
                           
                           # if hit on transcript
                           anno =&gt; {
                             $tid =&gt; {
                               gsym =&gt; $gene_symbol,
                               gid  =&gt; $gene_id,
                               strd =&gt; $strand,
                               cpos =&gt; $hgvs_position,
                               exin =&gt; $exon_intron_number,
                               regcod =&gt; $code_of_region,
                               regtyp =&gt; $type_of_region,
                             },
                             ...
                           },
                        },
                        ...
                      },
                      ...
                }</code></pre>

<h1 id="DATABASE-FORMAT">DATABASE FORMAT</h1>

<p>The Format of annotation database is listed as following:</p>

<h2 id="BED-FORMAT-DATABASE">BED FORMAT DATABASE</h2>

<pre><code>   Departed block with tag for annotation, Tag entries are separated by &quot;; &quot;,
   and Infos items in entry are separated by &quot;|&quot;.

   Each entry contains the follwing infomation:

   1. Acc.Ver
   2. GeneID
   3. Gene Symbol
   4. Strand
                 5&#39;=====|&gt;&gt;&gt;|[=============]|&gt;&gt;&gt;&gt;&gt;&gt;|[==========]|&gt;&gt;&gt;&gt;&gt;&gt;|[=============]|&gt;&gt;&gt;&gt;|==3&#39;
   5. BlockAttr  : PROM 5U2E D5U1 I5U1 A5U1 5U1 C1  DC1 IC1 AC1 C2E 3U1 D3U1 I3U1 A3U1 3U2E
   6. GenePartsSO: 167  204  163  447  164  204 316 163 191 164 316 205 163  448  164  448
   7. ExIn Num   :    . |EX1|      IVS1     |  EX2 |    IVS2    |  EX3 |    IVS3       |EX4E|
   8. nHGVS start for block before departing
   9. nHGVS end for block before departing
   10.cHGVS start for block before departing
   11.cHGVS end for block before departing
   12.Length for block before departing
   13.MismatchBlock :  $type,$gstart,$gstop,$gseq
                       ($gseq is in the strand of refseq, &#39;.&#39; for deletion)
   14.Primary Tag : Please see &quot;PRIMARY TAG ASSIGNMENT&quot;
   15.Offset to leftmost of non departing block.</code></pre>

<h2 id="TRANSCRIPT-FASTA-DATABASE">TRANSCRIPT FASTA DATABASE</h2>

<pre><code>   One-line sequence fasta file
   ============================
   Header format is: ( separate by &quot; &quot;, with &quot;.&quot; for unavailable value )

       &gt;rnaAcc.ver rnaLen gene protAcc.ver protLen cdsSta,cdsEnd tags [altStartCodons] [frameshiftSite]
   
    &quot;tags&quot; are a string of multiple properties separated by &quot;|&quot;, which are:

        &lt;alignStat&gt;[|altstart][|selenocysteine][|inseqStop][|polyATail]

    when &quot;tags&quot; contain &quot;altstart&quot;, alternative start codons will be list in 
    &quot;altStartCodons&quot;, separated by &quot;;&quot;</code></pre>

<h2 id="PRIMARY-TAG-ASSIGNMENT">PRIMARY TAG ASSIGNMENT</h2>

<pre><code>    Our database need to sort the refSeq record with same Acc, 
    or with same gene by the following rules:

    1. whether the transcript is on the primary assembly.
    2. whether the CDS is 3-codons (badcds will be put to the tail)
    3. concatenated CDS is longer
    4. concatenated Exon is longer
    5. union Exon with flank region is longer
    6. Chromosome ID number is smaller
    7. Position number on forward strand-chromosome is smaller

    if LRG_RefSeqGene file is used, then the primary tag &quot;Y&quot; will
    only assign to the reference standard transcripts&#39;s first 
    mapping entry. But for non-RefSeqGene gene/transcript,
    assign primary tag &quot;Y&quot; to the first record for same genes&#39; all
    transcripts, which is the same with no LRG_RefSeqGene file case.

    For multiple-mapping of a same transcript, add postfix &quot;-N&quot; 
    (1..n-1) to the other records in the order of sort.</code></pre>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<pre><code>    HGVS     :  http://www.hgvs.org/mutnomen/recs.html
    Mutalyzer:  https://mutalyzer.nl</code></pre>

<h1 id="AUTHOR">AUTHOR</h1>

<p>liutao &lt;liut@geneplus.org.cn&gt;</p>

<h1 id="COPYRIGHT-AND-LICENSE">COPYRIGHT AND LICENSE</h1>

<p>Please check LICENSE file for detail</p>


</body>

</html>


